### PR TITLE
210 economy demand ratio replace lossy half up rounding with an exact comparison

### LINKED ISSUE #210: ECONOMY demand ratio: replace lossy HALF_UP rounding with an exact comparison
# ECONOMY demand ratio: replace lossy HALF_UP rounding with an exact comparison

**Parent:** #200
**Related:** #204 (introduced the pure decision), #205 (wiring & cleanup)
**Severity:** Medium
**Type:** Behaviour change (TDD applies)

## Context

`decideEconomy` (in `core/domain/heating/HeatingDecision.kt`) decides ON/OFF by
rounding the demand ratio to scale 2 with `HALF_UP` **before** comparing to the
threshold:

```kotlin
val demandRatio = areas.count { it.requiresHeating() }.toBigDecimal()
    .divide(areas.size.toBigDecimal(), 2, RoundingMode.HALF_UP)
if (demandRatio >= threshold.value) HeaterCommand.ON else HeaterCommand.OFF
```

`Percentage` allows arbitrary precision in `0..1`, so the scale-2 rounding is a
**lossy step in an otherwise pure/exact decision** and can flip the result at the
boundary.

## Example

5 of 8 areas demand heat → true ratio `0.625`, which `HALF_UP` rounds to `0.63`.
With `threshold = 0.63`, `0.63 >= 0.63` yields **ON**, even though the real ratio
`0.625` is **below** the threshold.

## Proposed fix

Compare exactly via cross-multiplication, avoiding any rounding:

```kotlin
// ON iff count / size >= threshold  <=>  count >= threshold * size
val demanding = areas.count { it.requiresHeating() }.toBigDecimal()
if (demanding >= threshold.value * areas.size.toBigDecimal()) HeaterCommand.ON else HeaterCommand.OFF
```

## Why this is deferred from #204

#204 is a strict **byte-for-byte port** of `EconomyAreasSharedHeatingStrategyService`,
which performs the identical scale-2 `HALF_UP` round before its `>=` comparison.
Adopting the exact comparison **changes the decision at boundary thresholds**, so it
cannot land inside the port — it needs its own boundary test and a deliberate sign-off.
Best sequenced alongside or after the #205 wiring.

## Acceptance criteria

- [ ] `decideEconomy` no longer rounds the ratio; the threshold comparison is exact.
- [ ] A unit test pins the boundary case that the old rounding got wrong
      (e.g. 5/8 with `threshold = 0.63` → `OFF`).
- [ ] If the old strategy is still wired at that point, confirm/justify the resulting
      behaviour difference (the rounding was a latent quirk, not a contract).

## Notes

Originated from PR review of #204 (reviewer thread, `HeatingDecision.kt:35`).

### PR DESCRIPTION
**Pull request by AGRFesta**

### CODE REVIEW

File: core/src/main/kotlin/org/agrfesta/sh/api/core/application/usecases/EvaluateHeatingStateService.kt [hunk 1]
```diff
@@ -49,17 +69,84 @@ class EvaluateHeatingStateService(
...
-        }.getOrNull()?.map { areasFactory.createArea(it, devicesRegistry) } ?: return
+        }.getOrNull() ?: return
+        val decide = strategySelector.select()
         val currentTime = timeProvider.currentLocalTime()
-        areas.filterIsInstance<HeatableArea>()
+        areaDtos.mapNotNull { it.resolveHeatable(devicesRegistry) }
             .groupBy { it.heater }
             .forEach { (heater, areaList) ->
-                strategy.handleHeatingFor(heater, areaList, currentTime)
+                val heaterStatus = heater.getActuatorStatus().getOrElse { ActuatorStatus.UNDEFINED }
```
Comment (AGRFesta) [inline core/src/main/kotlin/org/agrfesta/sh/api/core/application/usecases/EvaluateHeatingStateService.kt]: 🤖 **Reviewer** — _severity: medium · confidence: high_

Observability regression: `heater.getActuatorStatus().getOrElse { ActuatorStatus.UNDEFINED }` swallows the failure silently. The strategy this refactor replaces logged it — `AbstractSharedHeatingAreasStrategyService.requiresHeatingInHysteresisRange` did `logger.warn("Unable to fetch Heater '${heater.uuid}' status")` on the `Left` before degrading to `temp <= target`. Every other failure in this file is logged (device fetch, area fetch, sensor readings, temperature settings), so degrading the heater status to UNDEFINED without a log is now the lone silent failure. Suggest `.onLeft { logger.warn("Unable to fetch status for heater '${heater.uuid}': $it") }.getOrElse { ActuatorStatus.UNDEFINED }`. Degrading-to-UNDEFINED is correctly tested, but the lost log is not — consider asserting it.

File: core/src/main/kotlin/org/agrfesta/sh/api/core/application/usecases/EvaluateHeatingStateService.kt [hunk 2]
```diff
@@ -49,17 +69,84 @@ class EvaluateHeatingStateService(
...
+                        areaId = area.areaId,
+                        currentTemperature = area.currentTemperature(),
+                        targetTemperature = targetTemperatureOf(area.areaId, currentTime),
+                        heaterStatus = heaterStatus
+                    )
+                }
+                val command = decide(snapshots)
+                logger.info("Heating decision for heater '${heater.uuid}': $command (snapshots: $snapshots)")
+                when (command) {
+                    HeaterCommand.ON -> heater.on()
```
Comment (AGRFesta) [inline core/src/main/kotlin/org/agrfesta/sh/api/core/application/usecases/EvaluateHeatingStateService.kt:94]: 🤖 **Reviewer** — _severity: low · confidence: medium_

`heater.on()` / `heater.off()` return `Either<ActuatorOperationFailure, _>` and the result is dropped, so a failed physical actuation is invisible. This is pre-existing (the old strategy also ignored it), so not a regression — but this block is being rewritten here, which is a natural point to add `.onLeft { logger.error("Failed to turn heater '${heater.uuid}' ON/OFF: $it") }`. Consistent with the file's logging discipline and with the separate Gemini comment on the original block (line 95).

**Comment by AGRFesta** [comment]
🤖 **Reviewer** — review summary (round 1, sha `92a1b4a27c38ba9d4d312124976e8ea1d44009e6`)

Reviewed the full PR diff (#216) against dev. The headline #210 fix is correct: decideEconomy compares count >= threshold * size exactly, the BigDecimal value-based comparison removes the lossy scale-2 HALF_UP rounding, and it is well covered by a boundary unit test plus a property test with an independent integer cross-multiplication oracle. The broader heating refactor (pure HeatingDecision/HeatableAreaSnapshot/HeaterCommand + HeatingStrategySelector, port relocation, new ArchUnit logging/application-independence rules) preserves the old COMFORT/ECONOMY semantics, and the SwitchBot hub provider fix (#203) is correct. Percentage stays on the 0..1 scale across config and decision, so no scale mismatch. Changelog is compliant (pure refactor #204-#207 needs no entry; #210/#203 are listed under Fixed). Two observability findings on EvaluateHeatingStateService, the strongest being a warn log dropped by the refactor. Note (process, not blocking): the PR title names only #210 while the branch bundles ~5 issues.

File: core/src/main/kotlin/org/agrfesta/sh/api/core/application/usecases/EvaluateHeatingStateService.kt (hunk 1 above)
Comment (AGRFesta) [reply core/src/main/kotlin/org/agrfesta/sh/api/core/application/usecases/EvaluateHeatingStateService.kt]: 👤 **Author** — accepted.

Confirmed against git history: the `warn` lived in `SharedHeatingAreasStrategyService.requiresHeatingInHysteresisRange` and was dropped when #205 moved the heater-status degrade into this service. Restored as a `warn` (tolerated degradation, matching the prior severity) before the `getOrElse { UNDEFINED }`:

```kotlin
val heaterStatus = heater.getActuatorStatus()
    .onLeft { logger.warn("Unable to fetch status for heater '${heater.uuid}': $it") }
    .getOrElse { ActuatorStatus.UNDEFINED }
```

On the suggested log assertion: deliberately skipped for now — the log is an added side-effect, the returned status and the decision are unchanged, so the existing degrade-to-UNDEFINED test still holds. Happy to pin it with a log-capture assertion if you'd prefer.

File: core/src/main/kotlin/org/agrfesta/sh/api/core/application/usecases/EvaluateHeatingStateService.kt (hunk 2 above)
Comment (AGRFesta) [reply core/src/main/kotlin/org/agrfesta/sh/api/core/application/usecases/EvaluateHeatingStateService.kt:94]: 👤 **Author** — accepted.

Agreed it's pre-existing and not a regression, but since the `when` block is rewritten here it's the natural place to align it with the file's logging discipline. Added `.onLeft { logger.error(...) }` on both actuations:

```kotlin
HeaterCommand.ON -> heater.on()
    .onLeft { logger.error("Failed to turn heater '${heater.uuid}' ON: $it") }
HeaterCommand.OFF -> heater.off()
    .onLeft { logger.error("Failed to turn heater '${heater.uuid}' OFF: $it") }
```

File: core/src/main/kotlin/org/agrfesta/sh/api/core/application/usecases/EvaluateHeatingStateService.kt [hunk 3]
```diff
@@ -49,17 +69,84 @@ class EvaluateHeatingStateService(
...
+                    HeaterCommand.OFF -> heater.off()
+                    HeaterCommand.NONE -> {}
+                }
```
[the remark this answers is not in the document]
Comment (AGRFesta) [reply core/src/main/kotlin/org/agrfesta/sh/api/core/application/usecases/EvaluateHeatingStateService.kt]: 👤 **Author** — thanks, addressed.

Both observability gaps are now covered (heater status fetch + `on()`/`off()` actuation). One intentional deviation from the snippet: the status-fetch failure logs at `warn`, not `error` — it's a tolerated degradation (falls back to `UNDEFINED`), consistent with the behaviour the previous strategy had; `error` is reserved for the physical actuation failures.

File: core/src/main/kotlin/org/agrfesta/sh/api/core/application/usecases/EvaluateHeatingStateService.kt (hunk 1 above)
Comment (AGRFesta) [reply core/src/main/kotlin/org/agrfesta/sh/api/core/application/usecases/EvaluateHeatingStateService.kt]: 🤖 **Reviewer**

Verified at d0d4d4e: `getActuatorStatus().onLeft { logger.warn("Unable to fetch status for heater '${heater.uuid}': $it") }.getOrElse { ActuatorStatus.UNDEFINED }` restores the warn the #205 refactor dropped, at the right severity (tolerated degradation). The lost log is back; the silent-failure regression is closed. On the skipped log-capture assertion: conceded — the returned status and decision are unchanged and already pinned by the degrade-to-UNDEFINED test, and a log line is a side-effect, not the core invariant. Fine to leave unasserted. Resolved.

File: core/src/main/kotlin/org/agrfesta/sh/api/core/application/usecases/EvaluateHeatingStateService.kt (hunk 2 above)
Comment (AGRFesta) [reply core/src/main/kotlin/org/agrfesta/sh/api/core/application/usecases/EvaluateHeatingStateService.kt:94]: 🤖 **Reviewer**

Verified at d0d4d4e: both actuation branches now log on failure — `heater.on().onLeft { logger.error("Failed to turn heater '${heater.uuid}' ON: $it") }` and the OFF counterpart. A failed physical actuation is no longer invisible, and `error` is the right severity for it, consistent with the file's logging discipline. Resolved.

**Comment by AGRFesta** [comment]
🤖 **Reviewer** — review summary (round 2, sha `d0d4d4e1bc6abb11f7c0ef02c25ec85297a93169`)

Re-review of the delta 92a1b4a..d0d4d4e (#216). Both round-1 observability findings on EvaluateHeatingStateService are addressed: the heater-status fetch now logs warn before degrading to UNDEFINED (restoring the log the #205 refactor dropped), and the on()/off() actuation failures now log error via onLeft. Severity split (warn for tolerated degradation, error for physical actuation) is the correct call and matches the prior strategy's behaviour. The remaining delta is a clean docs addition to TDD_WORKFLOW.md (PBT guidance) with no contract impact. No new issues. Both reviewer threads resolved; converging.
